### PR TITLE
CP-19454: Support Chinese and Janpanese mail alert 

### PR DESCRIPTION
**Pull request by makunterry**
1. Add 'mail-languages' processing in mail-alarm script
2. Introduce 3 JSON files for mail language resource

Signed-off-by: Kun Ma <kun.ma@citrix.com>

### CODE REVIEW

File: scripts/mail-languages/ja-JP.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			"subject":"[%s] %s Alarm: Filesystem nearly full on \"%s\"",
+			"body":"The filesystem usage on \"%s\" is at %.1f%%.\nThis alarm is set to be triggered when filesystem usage is more than %.1f%%.\n\n"
+		},
+		"dom0logfs_usage_alarm":{
+			"subject":"[%s] %s Alarm: Log partition nearly full on \"%s\"",
+			"body":"The log partition usage on \"%s\" is at %.1f%%.\nThis alarm is set to be triggered when log partition usage is more than %.1f%%.\n\n"
+		},
+		"dom0memory_usage_alarm":{
+			"subject":"[%s] %s Alarm: Dom0 memory demand is high on \"%s\"",
+			"body":"The memory required by the control domain on \"%s\" is about %.1f%% of its allocated memory.Occasional performance degradation can be expected when memory swapping is forced to happen.\nThis alarm is set to be triggered when the memory required by the control domain is above %.1f%% of its allocated memory.\n\n"
```
Comment (lindig) [inline scripts/mail-languages/ja-JP.json]: Space missing before "Occasional".

File: scripts/mail-languages/en-US.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+			"subject":"[%s] %s Alarm: Dom0 memory demand is high on \"%s\"",
+			"body":"The memory required by the control domain on \"%s\" is about %.1f%% of its allocated memory.Occasional performance degradation can be expected when memory swapping is forced to happen.\nThis alarm is set to be triggered when the memory required by the control domain is above %.1f%% of its allocated memory.\n\n"
+		},
+		"wlb_consultation_failure":{
+			"subject":"[%s] %s Alarm: Attempt to consult wlb for VM \"%s\" failed",
+			"body":"A workload balancing consultation for VM %s failed.\nThe operation was completed using the default algorithm instead of a workload balancing recommendation.\n\n"
+		},
+		"wlb_optimization_alert":{
+			"subject":"Workload Balancing Alert: Optimization alert from pool %s",
+			"body":"The Workload Balancing server has reported that pool %s is in need of optimization.\n%s is in optimization mode %s and is in a %s state.\n\n"
```
Comment (lindig) [inline scripts/mail-languages/en-US.json]: I think this should be `Workload Balancing Server`.

File: scripts/mail-languages/en-US.json [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+		"dom0logfs_usage_alarm":{
+			"subject":"[%s] %s Alarm: Log partition nearly full on \"%s\"",
+			"body":"The log partition usage on \"%s\" is at %.1f%%.\nThis alarm is set to be triggered when log partition usage is more than %.1f%%.\n\n"
+		},
+		"dom0memory_usage_alarm":{
+			"subject":"[%s] %s Alarm: Dom0 memory demand is high on \"%s\"",
+			"body":"The memory required by the control domain on \"%s\" is about %.1f%% of its allocated memory.Occasional performance degradation can be expected when memory swapping is forced to happen.\nThis alarm is set to be triggered when the memory required by the control domain is above %.1f%% of its allocated memory.\n\n"
+		},
+		"wlb_consultation_failure":{
+			"subject":"[%s] %s Alarm: Attempt to consult wlb for VM \"%s\" failed",
```
Comment (lindig) [inline scripts/mail-languages/en-US.json]: I think this should be `WLB`.

**Comment by makunterry** [comment]
@lindig -- Thanks for you reviewing. I have updated code change, please have a look.

**Comment by lindig** [comment]
I'm fine with this but would like somebody with more Python experience to review the Python code. I also wonder why the different languages are not kept in one JSON file with the language being a key for a dictionary. The ticket doesn't say anything about this.

**Comment by makunterry** [comment]
@lindig -- Thanks! The reason for using separate JSON files is for easily extend in future. And this is included in feature design page.

File: scripts/Makefile [hunk 4]
```diff
@@ -154,5 +154,10 @@ install:
 	$(IPROG) poweron/iLO.py $(DESTDIR)$(PLUGINDIR)/iLO.py
 	$(IPROG) poweron/DRAC.py $(DESTDIR)$(PLUGINDIR)/DRAC.py
 	$(IPROG) poweron/power-on.py $(DESTDIR)$(PLUGINDIR)/power-on-host
+# maillanguages
+	mkdir -p $(DESTDIR)/etc/xapi.d/maillanguages
```
Comment (mseri) [inline scripts/Makefile]: Shouldn't this be `mkdir -p $(DESTDIR)/etc/xapi.d/mail-languages`?

**Comment by makunterry** [comment]
@mseri -- That was indeed a mistake. Thanks for pointing it out.

File: scripts/mail-alarm [hunk 5]
```diff
@@ -99,17 +100,44 @@ def get_sender(other_config):
...
-class EmailTextGenerator:
-    pass
+class EmailTextGenerator(object):
+    def __init__(self, mail_language, mail_language_component):
+        self.mail_language = mail_language
+        self.mail_language_component = mail_language_component
+
+    def mail_err(self, key):
+        log_err("[%s] error in looking for (%s:%s)." % (self.__class__.__name__, self.mail_language_component, key))
+        log_err("Mail language pack: %s." % (self.mail_language))
```
Comment (mseri) [inline scripts/mail-alarm]: I would make it into a unique line:

```python
log_err("[%s] error raised by mail language pack %s while looking for (%s:%s)." 
        % (self.__class__.__name__, self.mail_language, self.mail_language_component, key))
```

Comment (makunterry) [inline scripts/mail-alarm]: I did it as you said, but I divided them now. But considering the second log will be long, so I would like to put it in second line.

File: scripts/mail-alarm [hunk 6]
```diff
@@ -99,17 +100,44 @@ def get_sender(other_config):
...
+
+    def mail_subject(self):
+        try:
+            return self.mail_language['mail-alarms'][self.mail_language_component]['subject']
+        except:
```
Comment (mseri) [inline scripts/mail-alarm]: I would prefer to catch the expected error here and fail if something unexpected is happening:
```python
    except KeyError:
        mail_err('subject')
        return None
    except Exception as err:
        log_err("[%s] unexpected failure: %s" % (self.__class__.__name__, str(err))
        raise  # return None if you prefer not to fail completely
```

File: scripts/mail-alarm [hunk 7]
```diff
@@ -99,17 +100,44 @@ def get_sender(other_config):
...
+            mail_err('subject')
+            return None
+
+    def mail_body(self):
+        try:
+            return self.mail_language['mail-alarms'][self.mail_language_component]['body']
+        except:
```
Comment (mseri) [inline scripts/mail-alarm]: As above

File: scripts/mail-alarm [hunk 8]
```diff
@@ -125,16 +153,24 @@ class CpuUsageAlarmETG(EmailTextGenerator):
...
+        body_str = self.mail_body()
+        if body_str is None:
+            body_str = 'CPU usage on %s "%s" has been on average %.1f%% for the last %d seconds.\n' \
             'This alarm is set to be triggered when CPU usage is more than %.1f%%.\n' \
             '\n' \
             'For Alarm Settings, please log into your %s Console and click on "%s"->\n' \
-            '"Properties"->"Alerts"\n' % \
+            '"Properties"->"Alerts"\n'
+
+        return \
```
Comment (mseri) [inline scripts/mail-alarm]: you can save a line and write

```python
return body_str % \
```

File: scripts/mail-alarm [hunk 9]
```diff
@@ -160,16 +197,24 @@ class NetworkUsageAlarmETG(EmailTextGenerator):
...
+        body_str = self.mail_body()
+        if body_str is None:
+            body_str = 'Network usage on %s "%s" has been on average %d B/s for the last %d seconds.\n' \
             'This alarm is set to be triggered when Network usage is more than %d B/s.\n' \
             '\n' \
             'For Alarm Settings, please log into your %s Console and click on "%s"->\n' \
-            '"Properties"->"Alerts"\n' % \
+            '"Properties"->"Alerts"\n'
+
+        return \
```
Comment (mseri) [inline scripts/mail-alarm]: As above, you can save a newline, the line here is short.

File: scripts/mail-alarm [hunk 10]
```diff
@@ -192,16 +238,24 @@ class MemoryUsageAlarmETG(EmailTextGenerator):
...
+        body_str = self.mail_body()
+        if body_str is None:
+            body_str = 'Free memory on %s "%s" has been on average %d KiB for the last %d seconds.\n' \
             'This alarm is set to be triggered when free memory is less than %d KiB.\n' \
             '\n' \
             'For Alarm Settings, please log into your %s Console and click on "Server"->\n' \
-            '"Properties"->"Alerts"\n' % \
+            '"Properties"->"Alerts"\n'
+
+        return \
```
Comment (mseri) [inline scripts/mail-alarm]: As above

File: scripts/mail-alarm [hunk 11]
```diff
@@ -277,20 +350,29 @@ class Dom0LogFSUsageAlarmETG(EmailTextGenerator):
...
-        return \
-            'The log partition usage on "%s" is at %.1f%%.\n' \
+        body_str = self.mail_body()
+        if body_str is None:
+            body_str = 'The log partition usage on "%s" is at %.1f%%.\n' \
             'This alarm is set to be triggered when log partition usage is more than %.1f%%.\n' \
-            '\n' % \
+            '\n'
+
+        return \
```
Comment (mseri) [inline scripts/mail-alarm]: This can all go in one line

File: scripts/mail-alarm [hunk 12]
```diff
@@ -302,73 +384,108 @@ class Dom0MemUsageAlarmETG(EmailTextGenerator):
...
-            'The memory required by the control domain on "%s" is about %.1f%% of its allocated memory.' \
+        body_str = self.mail_body()
+        if body_str is None:
+            body_str = 'The memory required by the control domain on "%s" is about %.1f%% of its allocated memory.' \
             'Occasional performance degradation can be expected when memory swapping is forced to happen.\n' \
             'This alarm is set to be triggered when the memory required by the control domain is above %.1f%% of its allocated memory.\n' \
-            '\n' % \
+            '\n'
+            
+        return \
```
Comment (mseri) [inline scripts/mail-alarm]: This can all go in one line

**Comment by makunterry** [comment]
@mseri -- Updated.

File: scripts/mail-alarm [hunk 13]
```diff
@@ -277,20 +350,29 @@ class Dom0LogFSUsageAlarmETG(EmailTextGenerator):
...
-            '\n' % \
+            '\n'
+
+        return \
+            body_str % \
             (self.params['name_label'],
              self.value * 100.0,
              self.alarm_trigger_level * 100.0)
 
 class Dom0MemUsageAlarmETG(EmailTextGenerator):
```
Comment (mseri) [inline scripts/mail-alarm]: This name is counter-intuitive imho, but it's safer to stick with the name we've got.

File: scripts/mail-alarm [hunk 14]
```diff
@@ -302,73 +384,108 @@ class Dom0MemUsageAlarmETG(EmailTextGenerator):
...
-        return \
-            'A workload balancing consultation for VM %s failed.\n' \
+        body_str = self.mail_body()
+        if body_str is None:
+            body_str = 'A workload balancing consultation for VM %s failed.\n' \
             'The operation was completed using the default algorithm instead of a workload balancing recommendation.\n' \
-            '\n' % \
+            '\n'
+
+        return \
```
Comment (mseri) [inline scripts/mail-alarm]: This can go all in one line

File: scripts/mail-alarm [hunk 15]
```diff
@@ -302,73 +384,108 @@ class Dom0MemUsageAlarmETG(EmailTextGenerator):
...
-        return \
-            'The Workload Balancing server has reported that pool %s is in need of optimization.\n' \
+        body_str = self.mail_body()
+        if body_str is None:
+            body_str = 'The Workload Balancing server has reported that pool %s is in need of optimization.\n' \
             '%s is in optimization mode %s and is in a %s state.\n' \
-            '\n' % \
+            '\n'
+
+        return \
```
Comment (mseri) [inline scripts/mail-alarm]: As above, this can go in one line

File: scripts/mail-alarm [hunk 16]
```diff
@@ -302,73 +384,108 @@ class Dom0MemUsageAlarmETG(EmailTextGenerator):
...
-            '%s\n' \
+        body_str = self.mail_body()
+        if body_str is None:
+            body_str = '%s\n' \
             '\n' \
             'This alarm is set to be triggered when a host belonging to a high availability pool fails.' \
-            '\n' % self.text
+            '\n'
+
+        return \
```
Comment (mseri) [inline scripts/mail-alarm]: All in one line

File: scripts/mail-alarm [hunk 17]
```diff
@@ -388,6 +505,12 @@ class XapiMessage:
         except:
             self.body      = ""
         self.pool_name = get_pool_name()
+        try:
+            with open('/etc/xapi.d/mail-languages/' + mail_language + '.json', 'r') as fileh:
+                self.ml = json.loads(fileh.read())
+        except:
```
Comment (mseri) [inline scripts/mail-alarm]: Catch the error and add it to the message. It could fail for many different reasons: file does not exists, file unreadable, encoding/decoding error, invalid json. We need to know why it is failing explicitly

File: scripts/mail-alarm [hunk 18]
```diff
@@ -388,6 +505,12 @@ class XapiMessage:
...
+            log_err("Read mail language pack [\"%s\"] error" % ('/etc/xapi.d/mail-languages/' + mail_language + ".json"))
+            self.ml = None
```
Comment (mseri) [inline scripts/mail-alarm]: I think `self.lang` or even `self.mail_language` is more explicative that `self.ml`

File: scripts/mail-alarm [hunk 19]
```diff
@@ -497,7 +620,8 @@ def main():
         min_priority = 3
 
     charset = other_config.get('mail-charset', 'utf-8')
-    msg = XapiMessage(sys.argv[1])
+    mail_language = get_mail_language(other_config)
+    msg = XapiMessage(sys.argv[1], mail_language)
```
Comment (mseri) [inline scripts/mail-alarm]: I know that it was already there, but I think we should check that `len(sys.argv) > 1` here and log an appropriate error if it is not the case

File: scripts/mail-alarm (hunk 5 above)
Comment (mseri) [inline scripts/mail-alarm]: I don't think it is long and find it easier to debug when the relevant informations come together, however if you feel strongly about it, I think we should at least add an header linking the two errors together. Something like

```python
log_err("[%s] raised by mail language pack %s." % (self.__class__.__name__, self.mail_language))
```

**Comment by makunterry** [comment]
@mseri -- Updated.

File: scripts/mail-alarm [hunk 20]
```diff
@@ -502,11 +502,19 @@ class XapiMessage:
             self.body      = ""
         self.pool_name = get_pool_name()
         try:
-            with open('/etc/xapi.d/mail-languages/' + mail_language + '.json', 'r') as fileh:
-                self.ml = json.loads(fileh.read())
+            mail_language_file = '/etc/xapi.d/mail-languages/' + mail_language + '.json'
+            with open(mail_language_file, 'r') as fileh:
+                file_content = fileh.read()
+                self.mail_language = json.loads(file_content, encoding="utf-8")
```
Comment (mseri) [inline scripts/mail-alarm]: This can be simplified into ```python
with open(mail_language_file, 'r') as fileh:
    self.mail_language = json.load(fileh, encoding="utf-8")
```

File: scripts/mail-alarm [hunk 21]
```diff
@@ -607,6 +615,10 @@ class XapiMessage:
 
 
 def main():
+    if len(sys.argv) < 2:
+        log_err("Wrong command with insufficient arguments:[\"%s\"]. Expected at least 1 argument\n" % (' '.join(sys.argv)))
```
Comment (mseri) [inline scripts/mail-alarm]: Great. I would reword it as followS:
```
log_err("Expected at least 1 argument but got %d: [\"%s\"]." % (len(sys.argv), ' '.join(sys.argv)))
```

File: scripts/mail-alarm (hunk 20 above)
Comment (makunterry) [inline scripts/mail-alarm]: I know your point. But what I think is that in this way we can easily print file content to log if any error occurs.

Comment (mseri) [inline scripts/mail-alarm]: But we are not printing the content, and in case we really need it (and for some reason we cannot access the file) we can always make the change temporarily in a dev rpm.

**Comment by makunterry** [comment]
@mseri -- Updated. And pushed force because it's a little code style change.

File: scripts/mail-alarm (hunk 17 above)
Comment (mseri) [inline scripts/mail-alarm]: Please add the actual error:

```python
except Exception as exn:
    log_err("Unknown Error from Decode [\"%s\"] to JSON object: \"%s\"" % (mail_language_file, str(exn)))
    self.mail_language = None
```

**Comment by mseri** [comment]
@makunterry only a small thing missing in `mail-alarm:514`. After you add the fix I am happy to accept and merge
